### DISCUSSION TITLE
Loading new theme using themes kitten resets font size

### DISCUSSION THREAD

**Discussion by mizlan** [kovidgoyal/kitty#6268] [Q&A] [answered]
Sometimes I like to switch my theme in the middle of a session. However, the theme kitten doesn't just load the new theme, it sends a signal to reload the entire config, which among other things sets my font size to the default again, which I do not often want (sometimes I set my font size a little larger for readability). In those cases I have to readjust my font size to the size it was before the theme switch.

I am wondering if there is a way to only load the current-theme.conf file, as opposed to the entire configuration. Or, alternatively, if there is a way to specify a font size for *new* kitty windows but not necessarily in the configuration directly itself, so that it is not triggered when the configuration is reloaded.

**Comment by kovidgoyal** [comment]
Use --dump-theme with the kitten and kitten @ set-colors (aka remote control to only change the colors).

**Reply by mizlan** [reply]
The way I currently have it set up is I need both the kitten @ set-colors AND the +kitten themes to get my colorscheme to fully change, with the *same theme file* being loaded in both cases. The @ set-colors is sufficient for everything except for some elements in the tab bar, which don't seem to get updated. I am looking into the source code to try to figure out what the +kitten themes does that @ set-colors does not, since copying my entire config into set-colors also doesn't fix the problem.

**Reply by mizlan** [reply]
I believe the following patch fixes my problem:

```diff
diff --git a/kitty/boss.py b/kitty/boss.py
index 860ed8072..bf5c81a89 100644
--- a/kitty/boss.py
+++ b/kitty/boss.py
@@ -2379,6 +2379,7 @@ def patch_colors(self, spec: Dict[str, Optional[int]], configured: bool = False)
             t = tm.active_tab
             if t is not None:
                 t.relayout_borders()
+            tm.apply_options()
         patch_global_colors(spec, configured)
 
     def apply_new_options(self, opts: Options) -> None:
```

(View properly here at my fork: https://github.com/kovidgoyal/kitty/commit/c61fcf943cfd9fbee23f631a138db7af8599b412)

I took this line from `load_config` which is called with +kitten themes, which I suspected may be necessary for refreshing some stuff within the tab bar. Let me know if this seems problematic in any way. I could also make a PR if you'd like.

**Comment by kovidgoyal** [comment]
apply_options is a bit too broad it can have unwanted side-effects. 
tab_bar.patch_colors() which is called by set_colors *should* be
updating everything. If you can tell me whats not being updated I can
take a look.

On Sat, May 13, 2023 at 06:01:46AM +0000, Michael Lan wrote:
> I believe the following patch fixes my problem:
> 
> ```diff
> diff --git a/kitty/boss.py b/kitty/boss.py
> index 860ed8072..bf5c81a89 100644
> --- a/kitty/boss.py
> +++ b/kitty/boss.py
> @@ -2379,6 +2379,7 @@ def patch_colors(self, spec: Dict[str, Optional[int]], configured: bool = False)
>              t = tm.active_tab
>              if t is not None:
>                  t.relayout_borders()
> +            tm.apply_options()
>          patch_global_colors(spec, configured)
>  
>      def apply_new_options(self, opts: Options) -> None:```
> 
> I took this line from `load_config` which is called with +kitten themes, which I suspected may be necessary for refreshing some stuff within the tab bar. Let me know if this seems problematic in any way. I could also make a PR if you'd like.
> 
> -- 
> Reply to this email directly or view it on GitHub:
> https://github.com/kovidgoyal/kitty/discussions/6268#discussioncomment-5890339
> You are receiving this because you commented.
> 
> Message ID: ***@***.***>

-- 
_____________________________________

Dr. Kovid Goyal 
https://www.kovidgoyal.net
https://calibre-ebook.com
_____________________________________

**Reply by mizlan** [reply]
Hmm I think things are a bit complicated here. I recently patched my fork of kitty to support using the 16 ANSI colors in the tab bar (using the format `fmt.fg__3` for example) using aixterm escape sequence (https://github.com/mizlan/kitty/commit/65302f99754338c5b29eda3321f5b660e78021c3), and _these_ colors are the ones that do not get updated on set-colors. I'm not sure if this patch isn't actually needed because there is another way to achieve the same goal, but it seems like the tab bar is not aware of the new 16 colors.

**Reply by mizlan** [reply]
I would appreciate it if you could guide me in the right direction, since understandably this is outside your responsibility because my patch changes the behavior of the program, but I'm curious as to what I might do to resolve this. (Also, if I should make a PR for the ANSI colors..)

**Comment by kovidgoyal** [comment, marked as answer]
Look at patch_colors in tab_bar.py you need to patch the full color
profile of the screen object not just the fg/bg colors.
